migration/idempotent/011-maintenance.sql

```diff
@@ -533,31 +552,74 @@ CREATE OR REPLACE FUNCTION prom_api.config_maintenance_jobs(number_jobs int, new
     VOLATILE
     SET search_path = pg_catalog, pg_temp
 AS $func$
+BEGIN
+    PERFORM prom_api.config_maintenance_jobs('metrics', 'retention', number_jobs, new_schedule_interval, new_config);
+    PERFORM prom_api.config_maintenance_jobs('traces', 'retention', number_jobs, new_schedule_interval, new_config);
+    PERFORM prom_api.config_maintenance_jobs('metrics', 'compression', number_jobs, new_schedule_interval, new_config);
+    RETURN TRUE;
+END
+$func$
+LANGUAGE PLPGSQL;
+--redundant given schema settings but extra caution for security definers
+REVOKE ALL ON FUNCTION prom_api.config_maintenance_jobs(int, interval, jsonb) FROM PUBLIC;
+GRANT EXECUTE ON FUNCTION prom_api.config_maintenance_jobs(int, interval, jsonb) TO prom_admin;
+COMMENT ON FUNCTION prom_api.config_maintenance_jobs(int, interval, jsonb)
+IS 'Configure the number of maintenance jobs run by the job scheduler, as well as their scheduled interval. Sets identical settings for all job types.';
+
+CREATE OR REPLACE FUNCTION prom_api.config_maintenance_jobs(signal _ps_catalog.signal_type,
+                                                            job_type _ps_catalog.job_type, 
+                                                            number_jobs int,
+                                                            new_schedule_interval interval, 
+                                                            new_config jsonb = NULL)
+    RETURNS BOOLEAN
+    --security definer to add jobs as the logged-in user
+    SECURITY DEFINER
+    VOLATILE
+    SET search_path = pg_catalog, pg_temp
+AS $func$
 DECLARE
   cnt int;
   log_verbose boolean;
+  final_config jsonb;
 BEGIN
     --check format of config
     log_verbose := coalesce(new_config->>'log_verbose', 'false')::boolean;
+    final_config := coalesce(new_config, '{}'::jsonb) || jsonb_build_object('signal', signal::text) || jsonb_build_object('type', job_type::text);
 
+    -- delete jobs with the old config style
     PERFORM public.delete_job(job_id)
     FROM timescaledb_information.jobs
-    WHERE proc_schema = '_prom_catalog' AND proc_name = 'execute_maintenance_job' AND (schedule_interval != new_schedule_interval OR new_config IS DISTINCT FROM config) ;
+    WHERE proc_schema = '_prom_catalog' 
+      AND proc_name = 'execute_maintenance_job' 
+      AND NOT coalesce(config, '{}'::jsonb) ?& ARRAY['signal', 'type'];
 
+    PERFORM public.delete_job(job_id)
+    FROM timescaledb_information.jobs
+    WHERE proc_schema = '_prom_catalog' 
+      AND proc_name = 'execute_maintenance_job' 
+      AND coalesce(config->>'signal', '') = signal::text
+      AND coalesce(config->>'type', '') = job_type::text
```

Comment:
i don't think these coalesces are necessary are they? I guess it doesn't hurt much either though since we aren't indexing on the config.

Reply:
I'd argue it's easier to reason about when `NULL` isn't involved